olog.py

```diff
@@ -61,8 +82,10 @@ async def aput_logbook(self, logbook):
             res = await api.put(f'logbooks/{logbook["name"]}',  json=logbook)
             logbook_from_server = res.json()
             logbook_from_server.pop('owner')
-            logbook.pop('owner')
-        if (not res.raise_for_status()) and (logbook != logbook_from_server):
+            logbook_cp = logbook.copy()
+            logbook_cp.pop('owner')
+        if (not res.raise_for_status()) and \
+                (logbook_cp != logbook_from_server):
             raise ValueError("No http error was raised but server doesn't \
```

Comment:
This error message strikes me as a little cryptic. Do we understand under what circumstance this would happen? If the server _knows_ that something went wrong, I would expect it to return a error status code of some kind. If, on the other hand, the server thinks everything went OK---that is, if it is just returning a copy of what it inserted into its database so we can verify it against the content of our requests---then I think any discrepancy we find indicates an uncaught _server_ error, not a user error. Since [`ValueError` is meant to communicate that the caller passed in inappropriate inputs](https://docs.python.org/3/library/exceptions.html#ValueError) I think that a better fit might be some custom exception class like `UncaughtServerError` subclassing `Exception`. It would also be helpful if the error message included the submitted `logbook` and the `logbook_from_server`, which would provide useful detail in a bug report to olog-es if and when this occurs.

Finally, when I suggested that we return `None` from all methods like this, I did not know that the response from the server actually contained a little extra information (the `'owner'` key); it isn't quite exactly a copy of what we submitted. Given that, it might be useful to return `res.json()` after all.

Reply:
One of the example is server always uses whoever login as the `owner` and overwrites user's submission dictionary.
```
cli = Client(URL, "Tom", PASSWORD)
cli.put({'owner': 'Bob', 'foo': 'bar'....} 
```
will end up have a ` {'owner': 'Tom', 'foo': 'bar'....}` in database. So verification could give us a chance to capture unexpected('owner' is expected in current behavior) diff between submission and response.

So just double check, below is what we want.
```
...
res.raise_for_status()
# The server returned OK. Its response contains a copy of what it
# inserted into its database. We can compare it with our submission for
# extra verification that everything worked correctly.

# Compare part like
if logbook_cp != logbook_from_server:
    raise UncaughtServerError(...)

return res.json()
```

Reply:
> So verification could give us a chance to capture unexpected('owner' is expected in current behavior) diff between submission and response.

If we know that the server ignores `owner` and always uses the current user, shouldn't we do that comparison before making the request? We can check that `owner`, if specified, matches `user` (which we should stash in `__init__`) and raise a `ValueError` if not before even bothering the server about it.

> So just double check, below is what we want.

Yes. And as a minor improvement on what I originally suggested, you might move `res.raise_for_status()` up to where you first get the response. There is no need to go through `res.json()` and `logbook_cp = ...` if the response code is an error.

Reply:
sounds good.